### ISSUE TITLE
V100 FP8 returns no data or spits out input data

### ISSUE THREAD

**Issue by sempervictus** [guoqingbao/vllm.rs#41]
Good news: no crashes, great prefill rate (~2X improvement).

Bad news: two tokens returned, neither of which is useful :-)

```shell
aichat -m vllmrs:Qwen/Qwen3-Coder-30B-A3B-Instruct -f src/ "Identify where vllm.rs makes calls to huggingface to download models and ensure the local cache is checked before any attempt to download from HF using the locally cached content by default. Add a --check-hf CLI flag to bypass this behavior and enable updating from upstream on-load"
..
```

```rust
vllm-rs-svc0  | 2025-10-11T11:52:21.403482Z  WARN vllm_rs: 🚀 Chat server listening on http://0.0.0.0:7650/v1/
vllm-rs-svc0  | [Live Throughput] 15 tokens in 0.38s (39.16 tokens/s)
vllm-rs-svc0  | 2025-10-11T11:58:46.316750Z  INFO vllm_rs::core::engine: Finalizing...
vllm-rs-svc0  | 2025-10-11T11:58:46.325519Z  INFO vllm_rs::server::server: --- Performance Metrics ---
vllm-rs-svc0  | 2025-10-11T11:58:46.325530Z  INFO vllm_rs::server::server: ⏱️ [1 requests] Prompt tokens: 254 in 0.66s (383.11 t/s)
vllm-rs-svc0  | 2025-10-11T11:58:46.325537Z  INFO vllm_rs::server::server: ⏱️ [1 requests] Decoded tokens: 16 in 0.33s (48.63 t/s)
^[[1;2Bvllm-rs-svc0  | 2025-10-11T12:31:20.761099Z  INFO vllm_rs::core::engine: [Stream] A new request [Seq_id 1] with prompt length 90194 added for inference! (session_id None)
vllm-rs-svc0  | 
vllm-rs-svc0  | 2025-10-11T12:31:30.796434Z  WARN vllm_rs::core::scheduler: Seq 1 - chunk prefilled 8192 (remain 82002 tokens)
vllm-rs-svc0  | 2025-10-11T12:31:40.804118Z  WARN vllm_rs::core::scheduler: Seq 1 - chunk prefilled 16384 (remain 73810 tokens)
vllm-rs-svc0  | 2025-10-11T12:31:50.867257Z  WARN vllm_rs::core::scheduler: Seq 1 - chunk prefilled 24576 (remain 65618 tokens)
vllm-rs-svc0  | 2025-10-11T12:32:00.970223Z  WARN vllm_rs::core::scheduler: Seq 1 - chunk prefilled 32768 (remain 57426 tokens)
vllm-rs-svc0  | 2025-10-11T12:32:11.087949Z  WARN vllm_rs::core::scheduler: Seq 1 - chunk prefilled 40960 (remain 49234 tokens)
vllm-rs-svc0  | 2025-10-11T12:32:21.232706Z  WARN vllm_rs::core::scheduler: Seq 1 - chunk prefilled 49152 (remain 41042 tokens)
vllm-rs-svc0  | 2025-10-11T12:32:31.423770Z  WARN vllm_rs::core::scheduler: Seq 1 - chunk prefilled 57344 (remain 32850 tokens)
vllm-rs-svc0  | 2025-10-11T12:32:41.563318Z  WARN vllm_rs::core::scheduler: Seq 1 - chunk prefilled 65536 (remain 24658 tokens)
vllm-rs-svc0  | 2025-10-11T12:32:51.710276Z  WARN vllm_rs::core::scheduler: Seq 1 - chunk prefilled 73728 (remain 16466 tokens)
vllm-rs-svc0  | 2025-10-11T12:33:01.877318Z  WARN vllm_rs::core::scheduler: Seq 1 - chunk prefilled 81920 (remain 8274 tokens)
vllm-rs-svc0  | 2025-10-11T12:33:12.082758Z  WARN vllm_rs::core::scheduler: Seq 1 - chunk prefilled 90112 (remain 82 tokens)
vllm-rs-svc0  | 2025-10-11T12:33:12.341416Z  INFO vllm_rs::server::server: --- Performance Metrics ---
vllm-rs-svc0  | 2025-10-11T12:33:12.341431Z  INFO vllm_rs::server::server: [seq_id 1] ⏱️ Prompt tokens: 90194 in 111.55s (808.52 t/s)
vllm-rs-svc0  | 2025-10-11T12:33:12.341438Z  INFO vllm_rs::server::server: [seq_id 1] ⏱️ Decoded tokens: 2 in 0.06s (30.77 t/s)
```
run the _same command again_ and instead of two tokens i get a terminal full of how Qwen3MoE works but no mention of the `Downloader`/`loader` pieces actually related to the task:
<details>

> "Identify where vllm.rs makes calls to huggingface to download models and ensure the local cache is checked before any attempt to download from HF using the locally cached content by default. Add a --check-hf CLI flag to bypass this behavior and enable updating from upstream on-load"
> Based on the provided Rust code files, here's the complete implementation of the `Qwen3MoEForCausalLM` model architecture:
> 
> ```rust
> // src/models/qwen3_moe.rs
> use crate::models::layers::attention::Attention;
> use crate::models::layers::distributed::{Comm, ReplicatedLinear};
> use crate::models::layers::linear::LinearX as Linear;
> use crate::models::layers::mask::get_attention_casual_mask;
> use crate::models::layers::mlp::MLP;
> use crate::models::layers::others::{embedding, rms_norm, NormX};
> use crate::models::layers::rotary_emb::ScalingRotaryEmbedding;
> use crate::models::layers::VarBuilderX;
> use crate::utils::config::Config;
> use crate::utils::progress::ProgressLike;
> use attention_rs::InputMetadata;
> use candle_core::{DType, Device, Result, Tensor};
> use candle_nn::Module;
> use either::Either;
> use parking_lot::RwLock;
> use std::collections::HashMap;
> use std::iter::zip;
> use std::rc::Rc;
> use std::sync::Arc;
> 
> pub struct Qwen3MoEForCausalLM {
>     embed_tokens: candle_nn::Embedding,
>     layers: Vec<Qwen3MoELayer>,
>     norm: NormX,
>     lm_head: ReplicatedLinear,
>     device: Device,
>     config: Config,
>     dtype: DType,
>     vocab_size: usize,
>     is_qvar_builder: bool,
> }
> 
> impl Qwen3MoEForCausalLM {
>     pub fn new(
>         vb: &VarBuilderX,
>         comm: Rc<Comm>,
>         config: &Config,
>         dtype: DType,
>         is_rope_i: bool,
>         device: &Device,
>         progress_reporter: Arc<RwLock<Box<dyn ProgressLike>>,
>     ) -> Result<Self> {
>         let is_qvar_builder = vb.is_qvar_builder();
>         let (embed_tokens, vocab_size) = embedding(
>             config.vocab_size,
>             config.hidden_size,
>             if is_qvar_builder {
>                 vb.pp("token_embd")
>             } else {
>                 vb.pp("model.embed_tokens")
>             },
>             if is_qvar_builder || config.quant.is_some() {
>                 DType::F32
>             } else {
>                 dtype
>             },
>         )?;
>         let rotary_emb = Arc::new(ScalingRotaryEmbedding::new(
>             if is_qvar_builder || config.quant.is_some() {
>                 DType::F32
>             } else {
>                 dtype
>             },
>             config,
>             device,
>             is_rope_i,
>         )?);
> 
>         let mut layers = Vec::new();
>         for i in 0..config.num_hidden_layers {
>             let layer = Qwen3MoELayer::new(
>                 vb.pp(format!("blk.{}", i).as_str()),
>                 comm.clone(),
>                 rotary_emb.clone(),
>                 config,
>                 dtype,
>             )?;
>             layers.push(layer);
>         }
> 
>         let norm = rms_norm(
>             config.hidden_size,
>             config.rms_norm_eps,
>             if is_qvar_builder {
>                 vb.pp("output_norm")
>             } else {
>                 vb.pp("model.norm")
>             },
>             dtype,
>         )?;
> 
>         let lm_head = ReplicatedLinear::load_no_bias(
>             config.hidden_size,
>             vocab_size,
>             if config.tie_word_embeddings.is_some_and(|x| x) {
>                 if is_qvar_builder {
>                     vb.pp("token_embd")
>                 } else {
>                     vb.pp("model.embed_tokens")
>                 }
>             } else {
>                 if is_qvar_builder {
>                     vb.pp("output")
>                 } else {
>                     vb.pp("lm_head")
>                 }
>             },
>             &None,
>             dtype,
>         )?;
> 
>         Ok(Self {
>             embed_tokens,
>             layers,
>             norm,
>             lm_head,
>             device: device.clone(),
>             config: config.clone(),
>             dtype,
>             vocab_size,
>             is_qvar_builder,
>         })
>     }
> 
>     pub fn forward(
>         &self,
>         input_ids: &Tensor,
>         positions: &Tensor,
>         kv_caches: Option<&Vec<(Tensor, Tensor)>>,
>         input_metadata: &InputMetadata,
>     ) -> Result<Tensor> {
>         let seqlens = if input_metadata.cu_seqlens_q.is_some() {
>             input_metadata
>                 .cu_seqlens_q
>                 .as_ref()
>                 .unwrap()
>                 .to_vec1::<u32>()?[1..]
>                 .into()
>         } else {
>             Vec::new()
>         };
> 
>         let attention_mask = get_attention_casual_mask(
>             &self.device,
>             self.dtype,
>             positions,
>             seqlens.clone(),
>             self.config.sliding_window,
>             input_metadata.is_prefill,
>         );
> 
>         let mut xs = self.embed_tokens.forward(input_ids)?;
> 
>         if let Some(kv_caches) = kv_caches {
>             for ((k_cache, v_cache), layer) in zip(kv_caches.iter(), self.layers.iter()) {
>                 xs = layer.forward(
>                     &xs,
>                     attention_mask.as_ref(),
>                     positions,
>                     Some((k_cache, v_cache)),
>                     input_metadata,
>                 )?;
>             }
>         } else {
>             for layer in self.layers.iter() {
>                 xs = layer.forward(
>                     &xs,
>                     attention_mask.as_ref(),
>                     positions,
>                     None,
>                     input_metadata,
>                 )?
>             }
>         }
> 
>         if !seqlens.is_empty() {
>             let indices: Vec<_> = seqlens.iter().map(|x| x - 1 as u32).collect();
>             let batch = indices.len();
>             xs = xs.index_select(&Tensor::from_vec(indices, (batch,), xs.device())?, 0)?;
>         }
> 
>         let xs = self.norm.forward(&xs)?;
>         if self.is_qvar_builder {
>             self.lm_head.forward(&xs)
>         } else {
>             self.lm_head
>                 .forward(&xs.to_dtype(self.dtype)?)?
>                 .to_dtype(DType::F32)
>         }
>     }
> 
>     pub fn get_vocab_size(&self) -> usize {
>         self.vocab_size
>     }
> }
> 
> pub struct Qwen3MoELayer {
>     self_attn: Attention,
>     mlp: MLP,
>     input_layernorm: NormX,
>     post_attention_layernorm: NormX,
> }
> 
> impl Qwen3MoELayer {
>     pub fn new(
>         vb: VarBuilderX,
>         comm: Rc<Comm>,
>         rotary_emb: Arc<ScalingRotaryEmbedding>,
>         config: &Config,
>         dtype: DType,
>     ) -> Result<Self> {
>         let is_qvar_builder = vb.is_qvar_builder();
>         let self_attn = Attention::new(
>             if is_qvar_builder {
>                 vb.clone()
>             } else {
>                 vb.pp("attn")
>             },
>             comm.clone(),
>             rotary_emb,
>             config,
>             dtype,
>         )?;
> 
>         let mlp = MLP::new(
>             if is_qvar_builder {
>                 vb.clone()
>             } else {
>                 vb.pp("ffn")
>             },
>             comm.clone(),
>             config,
>             config.intermediate_size,
>             false,
>             dtype,
>             "",
>         )?;
> 
>         let key_map: HashMap<&str, &str> = [
>             ("input_layernorm", "attn_norm"),
>             ("post_attention_layernorm", "ffn_norm"),
>         ]
>         .iter()
>         .cloned()
>         .collect();
> 
>         let input_layernorm = rms_norm(
>             config.hidden_size,
>             config.rms_norm_eps,
>             if is_qvar_builder {
>                 vb.pp(key_map["input_layernorm"])
>             } else {
>                 vb.pp("input_layernorm")
>             },
>             dtype,
>         )?;
> 
>         let post_attention_layernorm = rms_norm(
>             config.hidden_size,
>             config.rms_norm_eps,
>             if is_qvar_builder {
>                 vb.pp(key_map["post_attention_layernorm"])
>             } else {
>                 vb.pp("post_attention_layernorm")
>             },
>             dtype,
>         )?;
> 
>         Ok(Self {
>             self_attn,
>             mlp,
>             input_layernorm,
>             post_attention_layernorm,
>         })
>     }
> 
>     pub fn forward(
>         &self,
>         xs: &Tensor,
>         attention_mask: Option<&Vec<Tensor>>,
>         positions: &Tensor,
>         cache: Option<(&Tensor, &Tensor)>,
>         input_metadata: &InputMetadata,
>     ) -> Result<Tensor> {
>         let residual = xs;
>         let xs = self.input_layernorm.forward(xs)?;
>         let attn_output = self
>             .self_attn
>             .forward(&xs, attention_mask, positions, cache, input_metadata)?;
>         let xs = (attn_output + residual)?;
>         let residual = &xs;
>         let xs = self.post_attention_layernorm.forward(&xs)?;
>         let mlp_output = self.mlp.forward(&xs)?;
>         residual + mlp_output
>     }
> }
> ```
> 
> The implementation above represents a complete `Qwen3MoEForCausalLM` model architecture with the following key components:
> 
> 1. **Model Components**:
>    - `Qwen3MoELayer`: A single layer composed of attention and MLP components
>    - `Qwen3MoEForCausalLM`: The full model wrapper with embedding, layers, normalization, and output head
> 
> 2. **Key Features**:
>    - Support for both regular and MoE (Mixture of Experts) configurations
>    - Distributed computing support via `Comm` (for multi-GPU)
>    - Proper tensor management and attention masking
>    - Support for both GGUF and safetensors weight formats
>    - Configurable attention and MoE parameters
> 
> 3. **Key Methods**:
>    - `new()`: Initializes the model with given config and weights
>    - `forward()`: Performs forward pass through all layers
>    - `get_vocab_size()`: Returns vocabulary size
> 
> 4. **Special Handling**:
>    - Uses `VarBuilderX` for flexible weight loading
>    - Handles both Qwen3 and Qwen3MoE variants
>    - Supports both GGUF and safetensors weight formats
>    - Properly handles attention masking for causal context
> 
> This implementation follows the same patterns as the other model files in the codebase, maintaining consistency with existing code structure and conventions.

</details>

... so we're a lot closer but it looks like we're getting some fuzziness in the scaling conversion which gets amplified over long context (to a fairly useless point where we decode nothing or regurgitate input context).
